src/dynamic_programming/zero_matrix.md

```diff
@@ -0,0 +1,86 @@
+
+It is clear that the desired zero submatrix is bounded on all four sides by some ones, which prevent it from increasing in size and improving the answer. Therefore,  we will not miss the answer if we act as follows: first, we iterate over the `i` bottom row of the zero submatrix, then we iterate over which column `j` we will rest up on the zero submatrix. Using the value `d[i][j]`, we immediately get the number of the top line of the zero submatrix. It now remains to determine the optimal left and right boundaries of the zero submatrix, i.e. maximally push this submatrix to the left and right of the `j`-th column. 
+
+What does it mean to push the maximum to the left? It means to find an index `k1` for which `d[i][k1] > d[i][j]`, and at the same time `k1` - the closest one to the left of the index `j`. It is clear that then `k1 + 1` gives the number of the left column of the required zero submatrix. If there is no such index at all, then put `k1` = `-1`(this means that we were able to extend the current zero submatrix to the left all the way to the border of matrix `a`).
+
+
+
+Symmetrically, you can define an index `k2` for the right border: this is the closest index to the right of `j` such that `d[i][k2] > d[i][j]` (or `m`, if there is no such index).
+
+So, the indices `k1` and `k2`, if we learn to search for them effectively, will give us all the necessary information about the current zero submatrix. In particular, its area will be equal to `(i - d[i][j]) * (k2 - k1 - 1)`.
+
+How to look for these indexes `k1` and `k2` effectively with fixed `i` and `j`? We can do that in `O(1)` on average.
+
+To achieve such complexity, you can use the stack as follows. Let's first learn how to search for an index `k1`, and save its value for each index `j` within the current row `i` in matrix `d1[i][j]`. To do this, we will look through all the columns `j` from left to right, and we will store in the stack only those columns that have `d[][]` strictly greater than `d[i][j]`. It is clear that when moving from a column `j` to the next column, it is necessary to update the content of the stack. When there is an inappropriate element at the top of the stack (i.e. `d[][] <= d[i][j]`) pop it. It is easy to understand that it is enough to remove from the stack only from its top, and from none of its other places (because the stack will contain an increasing `d` sequence of columns).
+
+The value `d1[i][j]` for each `j` will be equal to the value lying at that moment on top of the stack.
+
+It is clear that since there are exactly `m` pieces added to the stack on each line, there could not be more deletions either, the sum of complexities will be linear.
+
+The dynamics `d2[i][j]` for finding the indices `k2` is considered similar, only you need to view the columns from right to left.
+
+It should also be noted that this algorithm consumes `O(m)` memory (not counting the input data - the matrix `a[][]`).
```

Comment:
Can you add the final time complexity O(n m)? The article only states the slower approach with O(n m^2).